certbot/tests/error_handler_test.py

```diff
@@ -26,34 +46,51 @@ def test_context_manager(self):
             with self.handler:
                 raise ValueError
         except ValueError:
-            pass
+            exception_raised = True
+
+        self.assertTrue(exception_raised)
         self.init_func.assert_called_once_with(*self.init_args,
                                                **self.init_kwargs)
 
-    @mock.patch('certbot.error_handler.os')
-    @mock.patch('certbot.error_handler.signal')
-    def test_signal_handler(self, mock_signal, mock_os):
-        # pylint: disable=protected-access
-        mock_signal.getsignal.return_value = signal.SIG_DFL
-        self.handler.set_signal_handlers()
-        signal_handler = self.handler._signal_handler
-        for signum in self.signals:
-            mock_signal.signal.assert_any_call(signum, signal_handler)
+    def test_context_manager_with_signal(self):
+        with signal_receiver(self.signals) as signals_received:
+            with self.handler:
+                should_be_42 = 42
+                send_signal(self.signals[0])
+                should_be_42 *= 10
 
-        signum = self.signals[0]
-        signal_handler(signum, None)
+        # check exectuion stoped when the signal was sent
+        self.assertEqual(42, should_be_42)
+        # assert signals were caught
+        self.assertEqual([self.signals[0]], signals_received)
+        # assert the error handling function was just called once
         self.init_func.assert_called_once_with(*self.init_args,
                                                **self.init_kwargs)
-        mock_os.kill.assert_called_once_with(mock_os.getpid(), signum)
-
-        self.handler.reset_signal_handlers()
         for signum in self.signals:
-            mock_signal.signal.assert_any_call(signum, signal.SIG_DFL)
+            sig = signal.getsignal(signum)
+            self.assertTrue((sig == signal.SIG_DFL) or (sig == signal.SIG_IGN))
```

Comment:
So I'm a bit worried about this. A lot of our package maintainers use our tests to make sure Certbot is working properly and signal handling can vary on different OSs. Will the signal always been `SIG_DFL` or `SIG_IGN` when running these tests? I have no idea.

With this in mind, I'd actually save the signal handlers in use at the start of this method and test that they've been restored to those values here. You could potentially do this by breaking some of the code from `signal_receiver` into a separate function that returns a `dict` of the signal handlers currently in use. The best way to do this in my opinion is to call the separate function and then loop over `signums` again to set the `lambda` signal handler.